pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -405,80 +414,111 @@ func (f *fqdnController) deleteRuleSelectedPods(ruleID string) error {
+			if _, exist := cachedDNSMeta.responseIPs[newIPStr]; !exist {
+				updateIPWithExpiration(newIPStr, newIPMeta)
 				addressUpdate = true
-				break
 			}
 		}
-		for oldIPStr, oldIP := range oldDNSMeta.responseIPs {
-			if _, ok := responseIPs[oldIPStr]; !ok {
-				if oldDNSMeta.expirationTime.Before(time.Now()) {
-					// This IP entry has already expired and not seen in the latest DNS response.
-					// It should be removed from the cache.
+
+		// check for presence of already cached IPs in the new response.
+		for cachedIPStr, cachedIPMeta := range cachedDNSMeta.responseIPs {
+			if newIPMeta, exist := newIPsWithExpiration[cachedIPStr]; !exist {
+				// The IP was not found in current response.
+				if cachedIPMeta.expirationTime.Before(currentTime) {
+					// this IP is expired and stale, remove it by not including it but also signal an update to syncRules.
 					addressUpdate = true
 				} else {
-					// Add the unexpired IP entry to responseIP and update the lowest applicable TTL if needed.
-					responseIPs[oldIPStr] = oldIP
-					if oldDNSMeta.expirationTime.Before(recordTTL) {
-						recordTTL = oldDNSMeta.expirationTime
-					}
+					// It hasn't expired yet, so just retain it with its existing expirationTime.
+					updateIPWithExpiration(cachedIPStr, cachedIPMeta)
 				}
+			} else {
+				// The cached IP is included in the current response; update its expiration time to the later of the new and existing values.
+				updateIPWithExpiration(cachedIPStr, ipWithExpiration{
+					ip:             cachedIPMeta.ip,
+					expirationTime: laterOf(newIPMeta.expirationTime, cachedIPMeta.expirationTime),
+				})
 			}
 		}
+
 	} else {
+		// This domain is being encountered for the first time.
+		// Check if it should be tracked by matching it against existing selectorItemToRuleIDs.
+
+		addToCache := false
 		for selectorItem := range f.selectorItemToRuleIDs {
 			// Only track the FQDN if there is at least one fqdnSelectorItem matching it.
 			if selectorItem.matches(fqdn) {
-				mustCacheResponse, addressUpdate = true, true
+				// A FQDN can have multiple selectorItems mapped, hence we do not break the loop upon a match, but
+				// keep iterating to create mapping of multiple selectorItems against same FQDN.
+				addToCache = true
 				f.setFQDNMatchSelector(fqdn, selectorItem)
 			}
 		}
+		if addToCache {
+			for ipStr, ipMeta := range newIPsWithExpiration {
+				updateIPWithExpiration(ipStr, ipMeta)
+			}
+			addressUpdate = true
+		}
 	}
-	if mustCacheResponse {
+
+	// ipWithExpirationMap remains empty only when FQDN doesn't match any selector.
+	if len(ipWithExpirationMap) > 0 {
 		f.dnsEntryCache[fqdn] = dnsMeta{
-			expirationTime: recordTTL,
-			responseIPs:    responseIPs,
+			responseIPs: ipWithExpirationMap,
 		}
-		f.dnsQueryQueue.AddAfter(fqdn, recordTTL.Sub(time.Now()))
 	}
+	if timeToRequery != nil {
+		// timeToRequery is never nil since ipWithExpirationMap is non-empty and updates when adding entries.
```

Comment:
Actually this comment seems wrong to me. At the time of last review, I think the `dnsQueryQueue.AddAfter` is conditioned around some checks similar to `if addToCache`. Now that it is not, timeToRequery can very much be `nil` if the fqdn does not need to be cached since no selectors is selecting it, and we do not return early in that case.
Reading Antonin's comment above, I do feel like this function is probably more complicated than it needs to be. What we should have is simply, in pseudo pseudo code
```
cachedDNSMeta, exist := f.dnsEntryCache[fqdn]
if !exist {
     if m := checkIfAnySelectorsMatchFqdn(); !m {
         // return early
      }
      addressUpdate = true
else {
      // We keep newIPsWithExpiration as the base reference since those are the data that are most likely already
      // accurate in terms of what we need to add to the dnsCache.
      // Loop thru data in cachedDNSMeta:
      //  If address is not in newIPsWithExpiration and expired, do nothing; simply ignore them
      //  If address is not in newIPsWithExpiration and not expired, add the exact entry into newIPsWithExpiration
      //  If address is in newIPsWithExpiration, set the exp time to laterOf(old, new)
      addressUpdate = keysMatch(cachedDNSMeta, newIPsWithExpiration) // fairly simple util
}
f.dnsEntryCache[fqdn] = dnsMeta{
       responseIPs: newIPsWithExpiration,
}
timeTorRequery := <earliest exp time of all IPs in newIPsWithExpiration>
```
This may not be the best performance-wise but I think it is okay. The more important thing is that logically it is much readable IMO and eliminates the possibility of some variables like `timeTorRequery` is unexpected nil but not easy to spot why

Reply:
The previous version of the code made sense to me:
```go
if len(ipWithExpirationMap) > 0 {
        // ...
        f.dnsQueryQueue.AddAfter(fqdn, timeToRequery.Sub(currentTime))
}
```
It is not possible for the map to not be empty while at the same time for `timeToRequery` to be `nil`. Both are updated together in `updateIPWithExpiration`.
"if the fqdn does not need to be cached since no selectors is selecting it" -> this is the one and only case where the `ipWithExpirationMap` map can be empty